### PR TITLE
Hidden stages

### PR DESCRIPTION
**Pull request by Bjvanminnen**
An earlier PR introduced some UI for hiding stages (hidden behind a flag), but didn't actually persist anything.

What is in this PR
(a) makes it so that levelbuilders can configure whether or not a script is allowed to have hideable stages and
(b) When a script is allowed to have hideable stages, and the client side flag is turned on, allows teachers to modify whether or not a stage is hidden, and persists this to the db.

What is not in this PR:
(a) We don't actually hide stages that are marked as hidden in the course overview page
(b) When you go to a level in a hidden stage we still show it.

Obviously there is still some work needed to finish this feature, but I wanted to keep my PRs smaller, and this seemed like a good stopping point.

I could also split the db migration into its own PR if needed.

### CODE REVIEW

File: apps/src/code-studio/hiddenStageRedux.js [hunk 1]
```diff
@@ -2,48 +2,107 @@
...
+export const UPDATE_HIDDEN_STAGES = 'hiddenStage/UPDATE_HIDDEN_STAGES';
+export const ALLOW_HIDEABLE = 'hiddenStage/ALLOW_HIDEABLE';
 
 export const hiddenStagesEnabled = () => experiments.isEnabled('hiddenStages');
 
+// State is a key indicating whether or not we've received hidden stage info
+// from the server, and then a set of stageIds mapped to bools that are true
+// if that stage is hidden
+const initialState = {
+  initialized: false
```
Comment (Bjvanminnen) [inline apps/src/code-studio/hiddenStageRedux.js:18]: I don't love this state construction, where it ends up looking something like:

```
state = {
   initialized: true,
   123: true,
   456: false
};
```

Better would probably be to add another level of nesting:

```
state = {
   initialized: true,
   stages: {
     123: true,
     456: false
  }
};
```

but this makes updates a lot more awkward.

Thoughts? I might be able to make the updates less awkward by using something like immutable.js

File: apps/src/code-studio/hiddenStageRedux.js [hunk 2]
```diff
@@ -2,48 +2,107 @@
...
+}
+
+
+export function getHiddenStages(scriptName) {
+  return (dispatch, getState) => {
+    if (!hiddenStagesEnabled()) {
+      return;
+    }
+
+    $.ajax({
```
Comment (Bjvanminnen) [inline apps/src/code-studio/hiddenStageRedux.js:89]: Could try to start using fetch, but in terms of getting the right polyfill and making sure this doesnt break browser compat, it seems like introducing fetch belongs in it's own PR at some point. Not sure when that point is.

File: apps/src/code-studio/progressRedux.js [hunk 3]
```diff
@@ -42,6 +42,8 @@ export default function reducer(state = initialState, action) {
       professionalLearningCourse: action.professionalLearningCourse,
       saveAnswersBeforeNavigation: action.saveAnswersBeforeNavigation,
       stages: action.stages.map(stage => _.omit(stage, 'hidden')),
+      // TODO - might make sense to split of some state that isnt actually specific to progress
+      scriptName: action.scriptName,
```
Comment (Bjvanminnen) [inline apps/src/code-studio/progressRedux.js]: I should remove this TODO, as I don't intend to address currently, but at some point it would be good to rationalize what info is stored in the progress subtree of our redux store.

**Comment by Bjvanminnen** [comment]
@islemaster to focus particularly on react/redux stuff @ashercodeorg to focus on migration and potentially other dashboard stuff

File: dashboard/app/controllers/script_levels_controller.rb [hunk 4]
```diff
@@ -90,6 +90,58 @@ def show
...
+  def toggle_hidden
+    section_id = params.require(:section_id).to_i
+    stage_id = params.require(:stage_id)
+    # this is "true" in tests but true in non-test requests
+    should_hide = params.require(:hidden) == "true" || params.require(:hidden) == true
+
+    section = Section.find(section_id)
+    authorize! :read, section
+
+    return head :forbidden unless Stage.find(stage_id).try(:script).try(:hideable_stages)
```
Comment (ashercodeorg) [inline dashboard/app/controllers/script_levels_controller.rb]: This is bad, in that it causes DB an unnecessary DB query (finding the stage should be cached). Please add a `TODO(asher)` to change this to use a cache.

File: dashboard/app/models/hidden_stage.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+#  index_hidden_stages_on_section_id  (section_id)
+#  index_hidden_stages_on_stage_id    (stage_id)
+#
+
+require 'cdo/section_helpers'
+
+class HiddenStage < ActiveRecord::Base
+  belongs_to :section
+  belongs_to :stage
+end
```
Comment (ashercodeorg) [inline dashboard/app/models/hidden_stage.rb]: Please separate the migration from the rest of the code (DTP'ing in separate DTPs).

Doing so assures nothing amiss happens should this migration or an earlier migration fail.

File: dashboard/app/models/hidden_stage.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+#  section_id :integer
+#  stage_id   :integer
+#
+# Indexes
+#
+#  index_hidden_stages_on_section_id  (section_id)
+#  index_hidden_stages_on_stage_id    (stage_id)
+#
+
+require 'cdo/section_helpers'
```
Comment (ashercodeorg) [inline dashboard/app/models/hidden_stage.rb]: OOC, why is this `require` required?

File: dashboard/app/models/script.rb [hunk 7]
```diff
@@ -79,7 +79,14 @@ def generate_plc_objects
...
 
-  serialized_attrs %w(pd admin_required professional_learning_course student_of_admin_required peer_reviews_to_complete)
+  serialized_attrs %w(
+    hideable_stages
+    pd
+    admin_required
+    professional_learning_course
+    student_of_admin_required
+    peer_reviews_to_complete
+  )
```
Comment (ashercodeorg) [inline dashboard/app/models/script.rb]: nit: Can these be alphabetized?

File: apps/src/code-studio/hiddenStageRedux.js (hunk 1 above)
Comment (islemaster) [reply apps/src/code-studio/hiddenStageRedux.js:18]: Do we have a guarantee that no stage will be called `initialized`? If so, this might be fine.  If you do want to try Immutable.js I think you'd use an immutable `Hash` and call `mergeDeep`?

File: dashboard/app/models/hidden_stage.rb (hunk 6 above)
Comment (Bjvanminnen) [reply dashboard/app/models/hidden_stage.rb]: I probably copy/pasted this file and accidentally left his in here? I can try without

File: dashboard/app/models/hidden_stage.rb (hunk 5 above)
Comment (Bjvanminnen) [reply dashboard/app/models/hidden_stage.rb]: Should it be just the migration + schema.rb change or also this new model file?

File: apps/src/code-studio/hiddenStageRedux.js (hunk 1 above)
Comment (Bjvanminnen) [reply apps/src/code-studio/hiddenStageRedux.js:18]: Stages are all numbers, so there shouldn't be any real risk of overlap.

File: apps/src/code-studio/hiddenStageRedux.js [hunk 8]
```diff
@@ -2,48 +2,107 @@
...
+      dataType: 'json',
+      contentType: 'application/json'
+    }).done(hiddenStageIds => {
+      // send updates for all stageIds, setting hidden to true
+      const updates = hiddenStageIds.reduce((obj, id) => {
+        return {
+          ...obj,
+          [id]: true
+        };
+      }, {});
```
Comment (islemaster) [inline apps/src/code-studio/hiddenStageRedux.js]: Nit: This is too clever for its own good; it initializes `hiddenStageIds.length + 1` objects when a reduce operation building a new object should only need one.

``` javascript
const updates = hiddenStageIds.reduce((updates, id) => {
  updates[id] = true;
  return updates;
}, {});
```

File: dashboard/app/models/hidden_stage.rb (hunk 5 above)
Comment (ashercodeorg) [reply dashboard/app/models/hidden_stage.rb]: Also this file, as otherwise anyone grabbing your commit and running `rake build` will generate this diff without it being part of their PR.

File: dashboard/app/models/hidden_stage.rb (hunk 6 above)
Comment (ashercodeorg) [reply dashboard/app/models/hidden_stage.rb]: This file should be auto-generated as a result of the migration? What copying and pasting is happening?

**Comment by islemaster** [comment]
Redux stuff looks good, tests look good. :+1:

File: dashboard/app/models/hidden_stage.rb (hunk 6 above)
Comment (Bjvanminnen) [reply dashboard/app/models/hidden_stage.rb]: Oh, probably none then. All I knew is I didn't remember adding this line :-P

File: dashboard/db/migrate/20160913194550_add_hidden_stages.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+class AddHiddenStages < ActiveRecord::Migration[5.0]
+  def change
+    create_table :hidden_stages do |hidden_stage|
```
Comment (ashercodeorg) [inline dashboard/db/migrate/20160913194550_add_hidden_stages.rb]: Naively, I wouldn't expect `hidden_stages` to be associated with `sections` from the name.

Does it make sense to name it `hidden_section_stages` or the like?

File: dashboard/test/controllers/script_levels_controller_test.rb [hunk 10]
```diff
@@ -1135,4 +1135,192 @@ def create_student_of_admin_script
...
+    hidden = JSON.parse(response.body)
+    assert_equal [], hidden
+
+    student = create :student
+    sign_in student
+
+    response = get_hidden(@script)
+    hidden = JSON.parse(response.body)
+    assert_equal [], hidden
+  end
```
Comment (ashercodeorg) [inline dashboard/test/controllers/script_levels_controller_test.rb]: nit: This should be two separate tests.

File: dashboard/test/controllers/script_levels_controller_test.rb [hunk 11]
```diff
@@ -1135,4 +1135,192 @@ def create_student_of_admin_script
...
+    section = put_student_in_section(student, teacher, @script)
+    stage1 = @script.stages[0]
+    assert !@script.hideable_stages
+
+    post :toggle_hidden, script_id: @script.id, stage_id: stage1.id, section_id: section.id, hidden: true
+    assert_response 403
+    assert_equal 0, HiddenStage.where(section_id: section.id).length
+  end
+
+  test "teacher can't hide or un  hide stages in sections they don't own" do
```
Comment (ashercodeorg) [inline dashboard/test/controllers/script_levels_controller_test.rb]: nit: unhide

File: dashboard/app/models/hidden_stage.rb (hunk 6 above)
Comment (Bjvanminnen) [reply dashboard/app/models/hidden_stage.rb]: The file is annotated as part of the migration, but does not appear to be generated that way. As part of my renaming this, I used `rails generate SectionHiddenStages` and didn't end up with the `require 'cdo/section_helpers'` leading me to believe that I had originally copy/pasted this file from elsewhere
